[Run] Fix Scientific Notation Errors in Calculator Plugin

viggyd opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

This PR allows the Calculator plugin in PowerToys Run to use scientific notation (e.g. `5e3 = 5000`). The PR fixes a regression in PowerToys that was introduced in v0.69.1.

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [X] **Closes:** #25656
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [X] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
Nothing else fixed. 

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
I updated the Calculator unit tests to include various expressions using this notation and confirmed that they all pass.

viggyd committed:
Parse scientific notation properly

After adding logic to replace `e` as a mathematical constant, there are bugs
when trying to use expressions like `5e3`. This change parses the
`<number>e<number>` format into expanded form to prevent replacement with
the constant.

Regex explanation: `(\d+\.*\d*)[eE](-*\d+)
`(\d+\.*\d*)`: Match any number of digits, followed by an optional `.` and
               more optional digits. The expression is used to capture things
               like: `5.0`, `1.`, `1` before the `e`
`[eE]`: Match either upper or lowercase `e`
`(-*\d+)`: Capture an optional `-` sign as well as a number of digits

viggyd committed:
Update regex to be more tolerant of weird entries

The new regex captures a wider variety of numbers. See this post for details
on the regex used:

https://stackoverflow.com/a/23872060

viggyd commented:
@microsoft-github-policy-service agree

viggyd committed:
Fix regular expression failing unit tests

Using `[` didn't capture the expression properly. Had to use `(` instead.

ekcom commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.Calculator/CalculateHelper.cs:
```diff
@@ -60,6 +61,13 @@ public static string FixHumanMultiplicationExpressions(string input)
 
+        private static string CheckScientificNotation(string input)
+        {
+            var p = @"(-?(\d+(\.\d*)?)|-?(\.\d+))[eE](-?\d+\.*\d*)";
```
```suggestion
            var p = @"(-?\d+(\.\d*)?)[eE](-?\d+(\.\d*)?)";
```
I've simplified the regex:
- Remove unnecessary capturing groups and consolidate mantissa section
- Fix the exponent section to be of the same form as the mantissa
  - Does not allow unlimited decimal points anymore

viggyd replied:
The simplified expression doesn't capture if anything that doesn't start with a digit, so if you had something like `.5e2`, it gets translated to `.(5 * 10^(2))`, which isn't right. That being said, if it's not valid to do something like `.5e2`, I like the new expression better

It's why that `|` is in there, since you could theoretically start with a decimal, though I do think that's unlikely

ekcom replied:
Oh, I see what you were doing now.
`2e.5` probably shouldn't work since exponential form is typically to a whole number, so that makes sense why you had the two sides different.

jaimecbernardo replied:
What do you mean with unlimited decimal points? What was being accepted before that now is not?

jaimecbernardo replied:
Ah, I see. The suggestions were declined.

viggyd replied:
The unlimited decimals is accepted because of the `\•*` after the `e`, should be `\.?` instead. Otherwise yeah, I wanted to leave the expression as is since not everyone likes to put digits before decimal places

ekcom commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.Calculator/CalculateHelper.cs:
```diff
@@ -60,6 +61,13 @@ public static string FixHumanMultiplicationExpressions(string input)
+        private static string CheckScientificNotation(string input)
+        {
+            var p = @"(-?(\d+(\.\d*)?)|-?(\.\d+))[eE](-?\d+\.*\d*)";
+            var r = "($1 * 10^($5))";
```
This line will need to be updated accordingly from the change in capturing groups.
```suggestion
            var r = "($1 * 10^($3))";
```

ekcom reviewed (approved):
Suggestions to improve/clarify regex included, but otherwise looks good.

Aradig26 reviewed (approved):
Please add these fast

jaimecbernardo reviewed (dismissed):
Hi, Sorry for the time I took in reviewing this PR. I thought changes were still being discussed.
Thanks for opening the PR.

It looks good to me, but I think scientific notation should only be with the upper case "E".
The main reason being that currently there can be some ambiguous expressions with the constant "e",

What do you think?

jaimecbernardo commented:
@viggyd , Please let me know what you think of the change I proposed (uppercase `E` only).
Trying to get this in before .75 and I'm closing out the merge window pretty soon.
If you're OK with the change, I can implement it as well.

viggyd commented:
> @viggyd , Please let me know what you think of the change I proposed (uppercase `E` only). Trying to get this in before .75 and I'm closing out the merge window pretty soon. If you're OK with the change, I can implement it as well.

Sorry, for some reason I wasn't getting notifications for this. There is one change in the regex to make because it could accept multiple decimal points.

I don't have strong opinions on the capital E, using lowercase is what I'm used to, but I agree it is ambiguous with other usage. I'll make it stricter to reduce that ambiguity. I can push the changes up this evening. If you need them sooner, you're welcome to make them also!

jaimecbernardo committed:
Allow only for uppercase E

jaimecbernardo commented:
> There is one change in the regex to make because it could accept multiple decimal points.

Thanks @viggy . I'm not sure what you mean with the multiple decimal points. Anyway, I don't think this is ready to go in yet after a re-review, because it doesn't support different decimal or number grouping characters.

viggyd commented:
> > There is one change in the regex to make because it could accept multiple decimal points.
> 
> Thanks @viggy . I'm not sure what you mean with the multiple decimal points. Anyway, I don't think this is ready to go in yet after a re-review, because it doesn't support different decimal or number grouping characters. I've added a commit to support only uppercase E and to add the expression to the the `RegValidExpressChar` so that upper case E is actually accepted.

Could you elaborate what you mean by different decimal or number grouping characters? I don't follow. I can add whatever additional changes are necessary

viggyd committed:
Only allow one decimal in second grouping

jaimecbernardo commented:
> Could you elaborate what you mean by different decimal or number grouping characters? I don't follow. I can add whatever additional changes are necessary

Hi @viggyd ,
There are regional formats where `.` is not the decimal point, but `,` is used instead. For example, "1.05" could be written like "1,05" on those countries.
![image](https://github.com/microsoft/PowerToys/assets/26118718/964b8e82-636d-4313-8364-98d645de28ed)

With these languages the calculator plugin supports the decimal point of the locale. Here's one example:
![image](https://github.com/microsoft/PowerToys/assets/26118718/6fbeb194-7f0c-401d-9b9c-dc1811fbeae6)

There's even languages that use . as a group separator. Here's Germany(German), for example:
![image](https://github.com/microsoft/PowerToys/assets/26118718/98b75217-73c0-457a-bc02-79aa962ce175)


But with this PR, it assumes . will always be the decimal separator:
![image](https://github.com/microsoft/PowerToys/assets/26118718/547d7215-0a4f-4182-9bbc-82ed3be870eb)
![image](https://github.com/microsoft/PowerToys/assets/26118718/4defa7a6-b895-495f-8181-d864f19d3aad)

viggyd commented:
> Hi @viggyd , There are regional formats where `.` is not the decimal point, but `,` is used instead. For example, "1.05" could be written like "1,05" on those countries. ![image](https://user-images.githubusercontent.com/26118718/277803475-964b8e82-636d-4313-8364-98d645de28ed.png)

Are there any functions to get the decimal separator?

jaimecbernardo commented:
> > Hi @viggyd , There are regional formats where `.` is not the decimal point, but `,` is used instead. For example, "1.05" could be written like "1,05" on those countries. ![image](https://user-images.githubusercontent.com/26118718/277803475-964b8e82-636d-4313-8364-98d645de28ed.png)
> 
> Are there any functions to get the decimal separator?

In NumberTranslator.cs there's references to `culture.NumberFormat.NumberGroupSeparator` and `culture.NumberFormat.NumberGroupSeparator`

htcfreek commented:
@viggyd 
Please check out the plugin settings for calculator as there is a setting to force English input format.

stefansjfw commented:
@viggyd Any progress here? Do you need some more help?

viggyd commented:
I'm sorry, I completely forgot about this issue. I'll be able to work on it this weekend, should have something that works by then

viggyd committed:
Support various decimal separator types

Previous regular expression did not allow decimal separators that were not ".".
The new expression uses the culture info decimal separator instead, which
should work better.

viggyd commented:
Okay, I pushed my update that accounts for various decimal separators, sorry for the delay!

viggyd committed:
Merge branch 'main' into module/run/fix-scientific-notation

Jay-o-Way commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.Calculator/CalculateHelper.cs:
```diff
@@ -60,6 +64,37 @@ public static string FixHumanMultiplicationExpressions(string input)
+            var p = @"(-?(\d+({0}\d*)?)|-?({0}\d+))E(-?\d+{0}?\d*)";
+
+            // If the decimal separator is a ".", we also have to insert a backslash.
+            if (CultureInfo.CurrentCulture.NumberFormat.NumberDecimalSeparator == ".")
+            {
+                p = string.Format(CultureInfo.CurrentCulture, p, @"\.");
+            }
+            else
+            {
+                p = string.Format(
+                    CultureInfo.CurrentCulture,
+                    p,
+                    CultureInfo.CurrentCulture.NumberFormat.NumberDecimalSeparator);
+            }
```
Technically, this can also be done with a ternary. But both would work.

viggyd replied:
I thought a ternary would take too much horizontal space because the `CultureInfo.CurrentCulture...` is a lot of characters so I left it at the expanded if/else instead

Jay-o-Way commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.Calculator/CalculateHelper.cs:
```diff
@@ -60,6 +64,37 @@ public static string FixHumanMultiplicationExpressions(string input)
+                    CultureInfo.CurrentCulture.NumberFormat.NumberDecimalSeparator);
+            }
+
+            var r = "($1 * 10^($5))";
```
Why create a variable for a single use?

viggyd replied:
No particular reason, I was just following a similar format to other functions that are already in the file. I updated the code to  remove the single use variable in my latest push

Jay-o-Way reviewed (commented):
Two details

stefansjfw commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.Calculator/CalculateHelper.cs:
```diff
@@ -60,6 +64,37 @@ public static string FixHumanMultiplicationExpressions(string input)
+             * E: Captures capital 'E'
+             * (-?\d+{0}?\d*): Captures decimal number starting with a number (e.g. "1.23")
+             */
+            var p = @"(-?(\d+({0}\d*)?)|-?({0}\d+))E(-?\d+{0}?\d*)";
```
can we have a decimal point after E? 🤔

stefansjfw replied:
i.e. should it be 
var p = @"(-?(\d+({0}\d*)?)|-?({0}\d+))E(-?\d+)";

Jay-o-Way replied:
Nope. The following number must be an integer.

viggyd replied:
Yeah, I went back and forth on this one and initially decided to allow users to do something like `1E1.2` if they really really wanted to do `1 * 10^(1.2)`, but it's definitely not the intended usage of the notation. I removed it so that it only allows integers after the `E` in my latest push

viggyd committed:
Merge branch 'main' into module/run/fix-scientific-notation

viggyd committed:
Only allow integers after `E`

viggyd committed:
Remove single use variable

stefansjfw commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.Calculator/CalculateHelper.cs:
```diff
@@ -18,6 +20,7 @@ public static class CalculateHelper
             @"sinh\s*\(|cosh\s*\(|tanh\s*\(|arsinh\s*\(|arcosh\s*\(|artanh\s*\(|" +
             @"pi|" +
             @"==|~=|&&|\|\||" +
+            @"((-?(\d+(\.\d*)?)|-?(\.\d+))[E](-?\d+\.*\d*))|" + /* expression from CheckScientificNotation between parenthesis */
```
should it be updated here as well? only integer after E

viggyd replied:
Whoops. Fixed!

stefansjfw replied:
thanks :) Change looks good to me, I'll just give it a test and then it'll be good to go! :)

viggyd committed:
Update regex to only accept integers after `E`

Missed this expression in my last update. Whoops

stefansjfw commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.Calculator/CalculateHelper.cs:
```diff
@@ -60,6 +64,36 @@ public static string FixHumanMultiplicationExpressions(string input)
+             * -?(\d+({0}\d*)?): Captures a decimal number starting with a number (e.g. "-1.23")
+             * -?({0}\d+): Captures a decimal number without leading number (e.g. ".23")
+             * E: Captures capital 'E'
+             * (-?\d+{0}?\d*): Captures an integer number (e.g. "-1" or "23")
```
we need a comment update here as well :D

stefansjfw replied:
fixed :)

viggyd replied:
Thanks!

stefansjfw commented on src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.Calculator/CalculateHelper.cs:
```diff
@@ -60,6 +64,36 @@ public static string FixHumanMultiplicationExpressions(string input)
+             * -?(\d+({0}\d*)?): Captures a decimal number starting with a number (e.g. "-1.23")
+             * -?({0}\d+): Captures a decimal number without leading number (e.g. ".23")
+             * E: Captures capital 'E'
+             * (-?\d+{0}?\d*): Captures an integer number (e.g. "-1" or "23")
```
```suggestion
             * (-?\d+): Captures an integer number (e.g. "-1" or "23")
```

stefansjfw committed:
Update src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.Calculator/CalculateHelper.cs

stefansjfw commented:
Hm, it still accepts and processes decimal numbers after E. And it gives weird results:
![image](https://github.com/microsoft/PowerToys/assets/57057282/dd0ddde0-ec0f-4bba-b51b-82083eda9c8b)

not sure why

stefansjfw commented:
> Hm, it still accepts and processes decimal numbers after E. And it gives weird results: ![image](https://private-user-images.githubusercontent.com/57057282/292927694-dd0ddde0-ec0f-4bba-b51b-82083eda9c8b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM2MzEyNjAsIm5iZiI6MTcwMzYzMDk2MCwicGF0aCI6Ii81NzA1NzI4Mi8yOTI5Mjc2OTQtZGQwZGRkZTAtZWMwZi00YmJhLWI1MWItODIwODNlZGE5YzhiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI2VDIyNDkyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwMDg1MTVkOTAwNmFiNGExZDdmN2QzMTZhODdlZWM1NWRjMmJjZTRmZWQxZWZjMWI4ODA0Yjg1Y2VmZGRlYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5RKZr9HXMPWHrQOFzGBDPWQP9Jg6Cic9QkWbdAuEDww)
> 
> not sure why

Actually it assumes multiplication between 1E1 and .3 -> 1E1 * 0.3 . Weird syntax and can confuse users.

viggyd commented:
Looks like that's just something you can do? Definitely weird behavior:
![image](https://github.com/microsoft/PowerToys/assets/38823133/45093be6-f2f2-4fd3-9835-cb81a2f339b1)

viggyd commented:
> > Hm, it still accepts and processes decimal numbers after E. And it gives weird results: ![image](https://private-user-images.githubusercontent.com/57057282/292927694-dd0ddde0-ec0f-4bba-b51b-82083eda9c8b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM2MzEyNjAsIm5iZiI6MTcwMzYzMDk2MCwicGF0aCI6Ii81NzA1NzI4Mi8yOTI5Mjc2OTQtZGQwZGRkZTAtZWMwZi00YmJhLWI1MWItODIwODNlZGE5YzhiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI2VDIyNDkyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwMDg1MTVkOTAwNmFiNGExZDdmN2QzMTZhODdlZWM1NWRjMmJjZTRmZWQxZWZjMWI4ODA0Yjg1Y2VmZGRlYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5RKZr9HXMPWHrQOFzGBDPWQP9Jg6Cic9QkWbdAuEDww)
> > not sure why
> 
> Actually it assumes multiplication between 1E1 and .3 -> 1E1 * 0.3 . Weird syntax and can confuse users.

Any ideas on how best to deal with that? Is there a way to reject a pattern?

viggyd commented:
Part of the problem is that there's a lot of other expressions that already allow expressions in parentheses followed by any number of spaces and it treats that as valid. So I can do something like `(10)    3` and I'll get `30` as a result even though it looks funny

stefansjfw commented:
> > > Hm, it still accepts and processes decimal numbers after E. And it gives weird results: ![image](https://private-user-images.githubusercontent.com/57057282/292927694-dd0ddde0-ec0f-4bba-b51b-82083eda9c8b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM2MzEyNjAsIm5iZiI6MTcwMzYzMDk2MCwicGF0aCI6Ii81NzA1NzI4Mi8yOTI5Mjc2OTQtZGQwZGRkZTAtZWMwZi00YmJhLWI1MWItODIwODNlZGE5YzhiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI2VDIyNDkyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwMDg1MTVkOTAwNmFiNGExZDdmN2QzMTZhODdlZWM1NWRjMmJjZTRmZWQxZWZjMWI4ODA0Yjg1Y2VmZGRlYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5RKZr9HXMPWHrQOFzGBDPWQP9Jg6Cic9QkWbdAuEDww)
> > > not sure why
> > 
> > 
> > Actually it assumes multiplication between 1E1 and .3 -> 1E1 * 0.3 . Weird syntax and can confuse users.
> 
> Any ideas on how best to deal with that? Is there a way to reject a pattern?

![image](https://github.com/microsoft/PowerToys/assets/57057282/887994be-f1bc-4660-8b9a-b5a93c6125fd)

looks like we accept this in general. I'm still worried that e.g. 1E1.5 might confuse users, because 10^1.5 is valid math expression, but E shouldn't accept decimal after (Windows calculator also prevent this when using exp).. Let's keep it as is as we respect the rules here.

there is another problem now. If I change the numbers format, as @jaimecbernardo mentioned above, to e.g. Portuguese:
![image](https://github.com/microsoft/PowerToys/assets/57057282/5dd45ff3-7216-40ee-b13a-5888c3d28335)

Regex is correctly matched, but the result is wrong 🤔

viggyd commented:
I see the problem. Ordinarily, the `NumberTranslator` converts the number/decimal separators so that by the time input reaches the calculator engine, it's already in English format (i.e. `.` for decimal separation). For some reason, the translator is having trouble parsing an expression like `2,3E2` and so it doesn't convert the expression to English format. The Mages engine PowerToys is using for calculating actual results requires English format and so it's not giving proper results.

That being said, I'm not sure why the translator isn't properly parsing the input. Seems like it's not recognizing the `E` as a token to split the expression on?

I can try to either:
1. Fix the translator so that the input we get when the user inputs scientific notation is in English and just remove the decimal separator logic in the calculator engine
2. Just make the calculator engine replace any decimal separators with the regular English separator. 

I think I'd prefer option 1 since that seems cleaner. I'll need to look into the translator and see how to make it recognize scientific notation input properly.

stefansjfw commented:
I agree. Let's try the option number  1

Jay-o-Way commented:
Yeah, those different possibilities for decimals are always a trouble. Option 2 actually sounds dangerous and could possibly lead to a regression resulting in bad calculations. Note that a number like  "one-thousand two-hundred thirty four" in some languages is written as `1,234` and in some others that would be `1.234`. In both cases, the other character denotes a decimal.

viggyd commented:
Okay, I think the issue in the NumberTranslator's `splitRegex` pattern. It's considering something like `2,3E2` to be a single token and it isn't splitting on the `E`:
![image](https://github.com/microsoft/PowerToys/assets/38823133/8e2dd2bb-4454-4a9c-ac7c-f92553c44062)

Because the regular expression is allowing characters `[a-fA-F]`, it's not creating a token split on the `E`. So the token as a whole (`2,3E2`) is sent to the `decimal.TryParse(...)`, which doesn't get translated to `2.3E2` presumably because the function doesn't understand the `E` in there. Interestingly, the same thing should happen for English decimal separators with the tokenization not working, but it works out there because there is no translation needed. 

I'm assuming the allowance of `a-fA-F` is to allow hex notation, but I didn't think PowerToys supported that. Anyone have insight into why these characters are allowed to be part of a singular token? If PowerToys does support hexadecimal notation now, I'm not sure how to fix the translator. Are we forcing all hex numbers to start with `0x`? If so, maybe I can look for that and split the token on `E` if there is no leading `0x`. Looks like @jaimecbernardo added the code hexadecimal code. Thoughts?

Jay-o-Way commented:
Darn it, you're absolutely right. In short, there is no way to tell if `5E3` is hexadecimal or scientific notation. Don't know how it would be possible to tell. Which jeopardizes the entire issue.

Having that said... Isn't scientific notation better to use for **results**, only? I mean, there's an inherent "insecurity" (or what's the word?) regarding significance. Like, when you say there are (about) 8 billion people on the planet, that's not an exact number. You could write it as 8B or 8E9. Note that both these are also valid as hexadecimal. What I mean is, you normally do not calculate anything with these kinds of estimates.

viggyd committed:
Update NumberTranslator to parse hex as a whole

viggyd committed:
Remove `hexRegex` as object member

The hex regex stuff is only used once, there's no need to keep track of it in
the object's state. Just create it during the translation process.

viggyd commented:
> Darn it, you're absolutely right. In short, there is no way to tell if `5E3` is hexadecimal or scientific notation. Don't know how it would be possible to tell. Which jeopardizes the entire issue. This is why Calculator app has modes for input.
> 
> Having that said... Isn't scientific notation better to use for **results**, only? I mean, there's an inherent "insecurity" (or what's the word?) regarding significance. Like, when you say there are (about) 8 billion people on the planet, that's not an exact number. You could write it as 8B or 8E9. Note that both these are also valid as hexadecimal. What I mean is, you normally do not calculate anything with these kinds of estimates.

I think maybe uncertainty is the word you're looking for?

While that certainly is a use case, I have used scientific notation in PowerToys in the past for the purposes of calculation. It's a convenience when working with smaller numbers, so that rather than needing to put in `0.001` I can just do `1e-3`. It's especially useful when converting between various SI units, dimensional analysis, etc.

It looks like PowerToys does require the `0x` prefix to denote a hexadecimal number. If I can assume that's always the case, I can do a 2-stage tokenization approach. The first can split out any hexadecimal expressions in the input. After that, all other tokens are free to be split without the `a-fA-F` in the splitting regex. The hexadecimals that I tokenized in the first stage can remain as one big group.

Something like (pseudocode):
```c#
string[] hex_norm_tokens = hexRegex.Split(input);

foreach (string token in hex_norm_tokens)
{
    if (token.startswith == "0x")
    {
         // Add to output builder - no further processing needed
         continue;
    }
    
    string[] reg_tokens = splitRegex.Split(token);
    // All current tokenization happens here
    ...
}
```

Should be able to split out hexadecimals on something like this: `(?:(0x[\da-fA-F]+)+)`. Regex101 link with test expression: [link](https://regex101.com/r/FZmpHj/1)

I tested this locally and its seems to work well. I'm wondering if I should update the unit tests as well. There's a few edge cases that we talked about during this conversation that'd be good to codify in unit tests somewhere to catch any regressions if they happen in the future

Jay-o-Way commented:
Alrighty. If the 0x prefix is/becomes mandatory, we must make sure this clear in the documentations. But if it works, it works.

viggyd committed:
Add unit tests for sci notation in other cultures

viggyd committed:
Merge branch 'main' into module/run/fix-scientific-notation

viggyd commented:
Okay, pushed a change that now includes a unit test for scientific notation in other cultures. It's very small though, let me know if there are any other cases I should test. I couldn't think of anything else that wasn't covered by a different test somewhere else. I can also move the tests I added in `QueryTests` into this new one I added under `ExtendedCalculatorParserTests`

stefansjfw reviewed (approved):
This looks good now! Nice work! Thanks for the contribution! :)
